pkg/collector/corechecks/system/windowscertificate/windows_certificate.go

```diff
@@ -0,0 +1,389 @@
+					"",
+					tags,
+					fmt.Sprintf("Certificate has expired. Certificate expiration date is %s", expirationDate))
+			} else if daysRemaining < float64(w.config.DaysCritical) {
+				sender.ServiceCheck("windows_certificate.cert_expiration",
+					servicecheck.ServiceCheckCritical,
+					"",
+					tags,
+					fmt.Sprintf("Certificate will expire in only %.2f days. Certificate expiration date is %s", daysRemaining, expirationDate))
+			} else if daysRemaining < float64(w.config.DaysWarning) {
+				sender.ServiceCheck("windows_certificate.cert_expiration",
+					servicecheck.ServiceCheckWarning,
+					"",
+					tags,
+					fmt.Sprintf("Certificate will expire in %.2f days. Certificate expiration date is %s", daysRemaining, expirationDate))
+			} else {
+				sender.ServiceCheck(
+					"windows_certificate.cert_expiration",
+					servicecheck.ServiceCheckOK,
+					"",
+					tags,
+					"",
+				)
+			}
+		}
+	}
+
+	return nil
+}
+
+func getCertificates(store string, certFilters []string) ([]*x509.Certificate, error) {
+	var certificates []*x509.Certificate
+	storeName := windows.StringToUTF16Ptr(store)
+
+	log.Debugf("Opening certificate store: %s", store)
+	storeHandle, err := windows.CertOpenStore(
+		windows.CERT_STORE_PROV_SYSTEM,
+		0,
+		0,
+		windows.CERT_SYSTEM_STORE_LOCAL_MACHINE|windows.CERT_STORE_READONLY_FLAG|windows.CERT_STORE_OPEN_EXISTING_FLAG,
+		uintptr(unsafe.Pointer(storeName)))
+	if err != nil {
+		log.Errorf("Error opening certificate store %s: %v", store, err)
+		return nil, err
+	}
+	// Close the store when the function returns
+	defer func() {
+		err = windows.CertCloseStore(storeHandle, 0)
+		if err != nil {
+			log.Errorf("Error closing certificate store %s: %v", store, err)
+		}
+	}()
+
+	log.Debugf("Enumerating certificates in store")
+
+	if len(certFilters) == 0 {
+		certificates, err = getEnumCertificatesInStore(storeHandle)
+	} else {
+		certificates, err = findCertificatesInStore(storeHandle, certFilters)
+	}
+	if err != nil {
+		log.Errorf("Error getting certificates: %v", err)
+		return nil, err
+	}
+
+	log.Debugf("Found %d certificates in store", len(certificates))
+	return certificates, nil
+}
+
+// getEnumCertificatesInStore retrieves all certificates in a certificate store
+func getEnumCertificatesInStore(storeHandle windows.Handle) ([]*x509.Certificate, error) {
+	var err error
+	certificates := []*x509.Certificate{}
+
+	var certContext *windows.CertContext
+	defer func() {
+		if certContext != nil {
+			log.Debugf("Freeing certificate context")
+			err = windows.CertFreeCertificateContext(certContext)
+			if err != nil {
+				log.Errorf("Error freeing certificate context: %v", err)
+			}
+		}
+	}()
+
+	for {
+		certContext, err = windows.CertEnumCertificatesInStore(storeHandle, certContext)
+		if err != nil {
+			if certContext == nil {
+				log.Debugf("No more certificates in store")
+				break
+			}
+			log.Errorf("Error enumerating certificates in store: %v", err)
+			return nil, err
+		}
```

Comment:
I do not think that this block is accurate. If you get an error will certContext be nil in all cases except complete enumeration? I am not sure but I feel that it could be nil in other cases as well. How Go maps native return code ERROR_NO_MORE_FILES and CRYPT_E_NOT_FOUND. They are "Ok" but everything else is a real error and require logging (instead of relying on context value that is). The same for the function below. I wonder if you should use an utility function which handles Cert retrival values and parsing since it is common for the code. Similarly, defer to close context probably can be to move to a dedicated function to make overall code smaller and more readable.